### PR TITLE
Fixed #25916 -- Added lastmod support to sitemapindex 

### PR DESCRIPTION
**Pull request by smithdc1**
[Ticket #25916](https://code.djangoproject.com/ticket/25916). This is my attempt at updating and addressing the review comments of https://github.com/django/django/pull/5814

I struggled with some of the documentation changes here. I'm really not sure if I've pitched it at the right level. 

The other main note is I introduced the `SiteData` to address backward compatibility concerns on the previous PR for custom indexes. I did wonder if this should be immediately deprecated?

### CODE REVIEW

File: django/contrib/sitemaps/views.py [hunk 1]
```diff
@@ -10,6 +10,15 @@
 from django.utils.http import http_date
 
 
+class SiteData:
+    def __init__(self, location, last_mod=None):
+        self.location = location
+        self.last_mod = last_mod
+
+    def __str__(self):
+        return self.location
```
Comment (carltongibson) [inline django/contrib/sitemaps/views.py:19]: We should probably keep this class. I think I'd call it `SitemapIndexItem`, and give it a docstring. 🤔 

I guess deprecating the `__str__()` usage will keep everyone on the same page. 

(Without something like this, we're still passing a list of pair or dicts or ... — given that we have the class already, we may as well keep it.)

**Review by carltongibson — COMMENTED** [review]
Hey @smithdc1 — thanks for this — sorry it's sat for so long. 

I think this looks good. I few comments but … 👍

File: tests/sitemaps_tests/test_generic.py [hunk 2]
```diff
@@ -85,3 +85,13 @@ def test_get_protocol_default_warning(self):
         )
         with self.assertWarnsMessage(RemovedInDjango50Warning, msg):
             sitemap.get_protocol()
+
+    def test_generic_sitemap_index(self):
+        TestModel.objects.first()
```
[the remark this answers is not in the document]
Comment (smithdc1) [reply tests/sitemaps_tests/test_generic.py]: Apparently I wrote this 10 months ago looking at the commit date. Which means I have no idea what so ever. (That's likely the case for much shorter timeframes too these days). 

Anyway, tests seem to pass without that which means I don't need to work it out again. 🎉

File: django/contrib/sitemaps/views.py (hunk 1 above)
Comment (smithdc1) [reply django/contrib/sitemaps/views.py]: I've changed this to a `dataclass`, what do you think about this? I don't think Django has any as yet? 

Following this change I'm stuggling for words on a doc string, well one that doesn't just say it's a dataclass to hold location and last_mod. What were you thinking of adding?

**Review by carltongibson — COMMENTED** [review]
Hey @smithdc1 — right, we're not far off here. 

Just a few more comments (again 😬 — sorry). 

Off-hand, can you recall why we don't call a callable `lastmod` by default? (No doubt there's a reason... 🙂) 

Thanks for the effort here. Let's get this in.

File: django/contrib/sitemaps/views.py (hunk 1 above)
Comment (carltongibson) [reply django/contrib/sitemaps/views.py]: I think it's just right as a dataclass 👍

**Pull request merged**